create-actor-screen-add-button

AsWiMoN opened the pull request:
- добавил кнопки DeleteButton, BackButton в новой ветке
- добавил Preview
- папка со скринами в корне проекта
[Кнопки](../blob/create-actor-screen-add-button/Screenshot/back_delete_buttons.png)

Commit:
Добавил Preview, добавил кнопки на активити добавил папку со скриншотами

ilhan76 commented:
Скрин просто в описание самого мра крепим, в проект не нужно

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/MainActivity.kt:
```diff
@@ -35,6 +36,7 @@ class MainActivity : ComponentActivity() {
                 Box(Modifier.background(color = UiKitTheme.colors.background)) {
                     CreateActorScreen()
                 }
+                ShowButton()
```
На Ui пока не добавляем элементы из кита, превью достаточно. ShowButton можно вообще удалить

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/screen/create_actor/items/backButton.kt:
```diff
@@ -0,0 +1,46 @@
+
+
+@Composable
+fun BackButton(modifier: Modifier = Modifier, onClick: () -> Unit) {
```
Этот композабл можно удалить, я в своем мре добавил его в качестве примера

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/screen/create_actor/items/deleteButton.kt:
```diff
@@ -0,0 +1,48 @@
+package ru.kudashov.rollinitiative.screen.create_actor.items
```
Перенеси в пакет ui.element.button.regular

AsWiMoN replied:
✔

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/screen/create_actor/items/deleteButton.kt:
```diff
@@ -0,0 +1,48 @@
+package ru.kudashov.rollinitiative.screen.create_actor.items
+
+import android.icu.text.ListFormatter.Width
```
Лишний импорт

AsWiMoN replied:
если этот импорт убираю в 14,15 строках появляется ошибка в этой строке

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/screen/create_actor/items/deleteButton.kt:
```diff
@@ -0,0 +1,48 @@
+import androidx.compose.ui.unit.sp
+
+@Composable
+fun DeleteButton(modifier: Modifier = Modifier, onClick: () -> Unit, width: Dp = 270.dp, height: Dp = 56.dp) {
```
Сделай каждый параметр с новой строки, пж

ilhan76 replied:
Размеры и стили сделай через appearance / size, по аналогии с моим пром

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/screen/create_actor/items/deleteButton.kt:
```diff
@@ -0,0 +1,48 @@
+
+@Composable
+fun DeleteButton(modifier: Modifier = Modifier, onClick: () -> Unit, width: Dp = 270.dp, height: Dp = 56.dp) {
+    Button(
```
Стараетмя максимально не использовать материаловский композаблы и пишем свои. Зачастую они плохо кастомизируются и имеют много лишнего. Здесь можно через Box сделать.

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/screen/create_actor/items/deleteButton.kt:
```diff
@@ -0,0 +1,48 @@
+    Button(
+        onClick = onClick,
+        colors = ButtonDefaults.buttonColors(
+            containerColor = Color(0xFFF5E6C8)
```
Цвета и шрифты используем из UiKitTheme

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/screen/create_actor/items/deleteButton.kt:
```diff
@@ -0,0 +1,48 @@
+            .height(height)
+    ) {
+        Text(
+            text = "Удалить",
```
Сроки выносим в xml ресурсы и достаем их через stringResource

ilhan76 replied:
И конкретно тут это вообще нужно вынести в параметр функции, потому что текст у кнопки может быть разный

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/screen/create_actor/items/deleteButton.kt:
```diff
@@ -0,0 +1,48 @@
+        Text(
+            text = "Удалить",
+            fontSize = 18.sp,
+            fontWeight = FontWeight.Bold,
```
Аналогично, шрифт и цвета достаем из темы. Но напрямую прям тут не прописываем, а выносим в size / appearance. 

Появняю, зачем это нужно. 

- В теме цвета будут автоматически менятся под нужную тему (спасибо, кэп), поэтому если мы что-то укажем напрямую, то при смене темы ничего не произойдет
- appearance / size упрощают работу с однотипными элементами. Например, кнопка у нас имеет несколько вариантов с разными размерами, цветами, скруглением углов и тд. Но везде это просто прямоугольник с текстом. И нет смысла его каждый раз верстать заново, если можно сделать его настраиваемым элементам и бесконечно расширять

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/screen/create_actor/items/deleteButton.kt:
```diff
@@ -0,0 +1,48 @@
+@Preview(showBackground = true)
+@Composable
+fun DeleteButtonPreview(){
+    DeleteButton(
```
В превью оборачиваем в тему

```kotlin
UiKitTheme {
    // your code
}
```

Это нужно для того, чтобы корректно определялись цвета и типографика из темы. Так же в превью следует добавлять динамически параметры через PreviewParameterProvider, если мы используем appearance / size (в остальных случаях это избыточно). И в таком случае туда так же нужно провайдить флаг темы. Если же PreviewParameterProvider не используется, то можно просто две превью функции сделать под каждую тему.

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/screen/create_actor/items/deleteButton.kt:
```diff
@@ -0,0 +1,48 @@
+@Composable
+fun DeleteButtonPreview(){
+    DeleteButton(
+        onClick = {}
```
Обработчики кликом можно сделать параметрами по умолчанию, чтобы не приходилось вот так вот заглушки передавать

ilhan76 commented:
Еще по поводу названия файлов с композаблами, называем из точно так же, как композабл, который там лежит. Например, IconButton лежит в файле IconButton.kt

ilhan76 commented:
**Рубрика "Полезные советы"**

Чтобы протестировать композабл, не обязательно добавлять его на экран, можно сбилдить превью на эмулятор, нажав на кнопку рядом с превью 

<img width="385" alt="Screenshot 2025-03-14 at 14 32 16" src="https://github.com/user-attachments/assets/3880f5b9-8811-4521-9ac5-52771036f60d" />

Или через контекстное меню на самом превью

<img width="434" alt="Screenshot 2025-03-14 at 14 32 36" src="https://github.com/user-attachments/assets/54ce3ebe-afa5-4795-87ab-419bdad85e5c" />

Commit:
Удалил backButton.kt, ShowButton.kt, попытался исправить все по примеру, с текстом пока что ничего не делал

Commit:
Исправил ошибку ссылки на NULL

Commit:
Добавил вызов текста из ресурсов в параметре функции, добавил вызов из appereance/size стиль текста, насколько я понял отрезолвил все замечания

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/ButtonStyleApearance.kt:
```diff
@@ -0,0 +1,64 @@
+    val Medium = ButtonSizeMedium()
+}
+
+interface IntButtonSize : UiKitSize, IntButtonAppearance {
```
Здесь нужно наследоваться только от uiKitSize, не объединяем два интерфейса в один. 

Еще такой момент, appearance / size именуем по такому формату {название композабла}Appearance/Size. То есть здесь должно быть так

- RegularButtonAppearance
- RegularButtonSize

Конкретные реализации по такому же принципу, только добавляем еще конкретики, например 
- RegularButtonSizeSmall
- RegularButtonAppearancePrimary

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/ButtonStyleApearance.kt:
```diff
@@ -0,0 +1,64 @@
+import ru.kudashov.rollinitiative.ui.element.base.UiKitSize
+import ru.kudashov.rollinitiative.ui.theme.UiKitTheme
+
+object ButtonStyleAppearances{
```
Для Size нужен такой же object

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/ButtonStyleApearance.kt:
```diff
@@ -0,0 +1,64 @@
+        TextStyle (
+            fontSize = 9.sp,
+            fontWeight = FontWeight.Bold,
+            color = Color(0xFF3E332A)
```
Цвет вынести в отдельное поле и передавать в Text(color = ...). Так же, не используем цвета напрямую, тянем только из темы. Для Текста это будет UiKitTheme.colors.primary

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/ButtonStyleApearance.kt:
```diff
@@ -0,0 +1,64 @@
+    override val size: @Composable () -> DpSize = { DpSize(width = 135.dp, height = 28.dp)},
+    override val roundedCornerShape: @Composable () -> Dp = { 8.dp },
+    override val textStyle: @Composable () -> TextStyle = {
+        TextStyle (
```
TextStyle тоже не используем напрямую, а тянем из темы – `UiKitTheme.typography`. Если там не нашлось подходящего, то создаем его там. Еще у тебя тут неправильный размер шрифта, чтобы посмотреть нужный, кликаешь в фигме на текст с зажатым alt и смотрришь справа характеристики текста. Размеры, цвет и тд, все там

<img width="578" alt="Screenshot 2025-03-18 at 14 16 51" src="https://github.com/user-attachments/assets/c60cfbb1-0456-4af0-9771-635d6242c13c" />

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/ButtonStyleApearance.kt:
```diff
@@ -0,0 +1,64 @@
+
+interface IntButtonSize : UiKitSize, IntButtonAppearance {
+    val size: @Composable () -> DpSize
+    val roundedCornerShape: @Composable () -> Dp
```
Два момента

- Не привязываемся названием переменной к реализации, то есть лучше назвать `backgroundShape`
- Лучше возвращать сразу Shape – `val backgroundShape: @Composable () -> Shape`

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/ButtonStyleApearance.kt:
```diff
@@ -0,0 +1,64 @@
+            color = Color(0xFF3E332A)
+        )
+    }
+) : IntButtonSize, IntButtonAppearance
```
Как уже писал, для каждого интерфейса своя реализация

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/deleteButton.kt:
```diff
@@ -0,0 +1,83 @@
+@Composable
+fun DeleteButton(
+    modifier: Modifier = Modifier,
+    appearance: IntButtonAppearance,
```
Тоже нужно значение по умолчанию

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/deleteButton.kt:
```diff
@@ -0,0 +1,83 @@
+    appearance: IntButtonAppearance,
+    size: IntButtonSize = ButtonStyleAppearances.Medium,
+    isEnabled: Boolean = true,
+    text: String = stringResource(id = R.string.delete_button),
```
А вот для текста значения по-умолчанию не нужно, это обязательное поле

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/deleteButton.kt:
```diff
@@ -0,0 +1,83 @@
+import ru.kudashov.rollinitiative.ui.element.button.regular.preview.DeleteButtonPreviewParameterProvider
+
+@Composable
+fun DeleteButton(
```
RegularButton

ilhan76 replied:
И файл все еще называется не так же, как композабл)

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/deleteButton.kt:
```diff
@@ -0,0 +1,83 @@
+    Box(
+        modifier = modifier
+            .size(size = size.size())
+            .clip(RoundedCornerShape((cornerRadius)))
```
Лучше обрезать через background

```
.background(
                color = appearance.backgroundColor(isPressed),
                shape = size.backgroundShape()
            )
```

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/deleteButton.kt:
```diff
@@ -0,0 +1,83 @@
+    size: IntButtonSize = ButtonStyleAppearances.Medium,
+    isEnabled: Boolean = true,
+    text: String = stringResource(id = R.string.delete_button),
+    onClick: () -> Unit,
```
Нужно значение по умолчанию.

Еще таком момент по порядку параметров в функции:

- Сначала идую обязательные поля
- Потом Modifier
- Потом остальные необязательные поля
- Последним идет лямбда для обработки кликов

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/deleteButton.kt:
```diff
@@ -0,0 +1,83 @@
+    ) {
+        Text(
+            text = text,
+            style = textStyle,
```
На всякий случай продублирую, цвет текста задаем здесь

ilhan76 replied:
Цвет не задан

ilhan76 replied:
Цвет текста лучше вынести в apperance

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/preview/DeleteButtonPreviewInfo.kt:
```diff
@@ -0,0 +1,10 @@
+
+import ru.kudashov.rollinitiative.ui.element.button.regular.IntButtonSize
+
+internal data class DeleteButtonPreviewInfo(
```
RegularButtonPreviewInfo

ilhan76 replied:
И текст параметром добавить

ilhan76 replied:
Текст не добавил

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/preview/DeleteButtonPreviewParameterProvider.kt:
```diff
@@ -0,0 +1,26 @@
+import androidx.compose.ui.tooling.preview.PreviewParameterProvider
+import ru.kudashov.rollinitiative.ui.element.button.regular.ButtonStyleAppearances
+
+internal class DeleteButtonPreviewParameterProvider : PreviewParameterProvider<DeleteButtonPreviewInfo> {
```
RegularButtonPreviewParameterProvider

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/preview/DeleteButtonPreviewParameterProvider.kt:
```diff
@@ -0,0 +1,26 @@
+                                size = size,
+                                appearance = appearance,
+                                useDarkMode = true,
+                                )
```
Коммент не только к этому файлу, почаще пользуйся ctrl + shift + L для форматирования кода

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/ButtonStyleApearance.kt:
```diff
@@ -0,0 +1,64 @@
+}
+
+interface IntButtonSize : UiKitSize, IntButtonAppearance {
+    val size: @Composable () -> DpSize
```
Посмотрел дизайн, у нас всегда кнопки растягиваются на максимальную ширину, поэтому в Size можно сделать параметр height, а ширину сделать .fillMaxWidth()

ilhan76 replied:
После того, как указал fillMaxWidth нужно уже не size указывать, а ширину. В данном случае fillMaxWidth, конечно, оказывается приоритетнее заданной, потому что идет перед size, но лучше таких конфликтов не допускать, потому что порядок модификаторов имеет значение

ilhan76 replied:
То есть у тебя должно быть так

```
Box(
    modifier = modifier
        .fillMaxWidth()
        .height(size.height())
        ...
)
    
interface RegularButtonSize : UiKitSize {
    val height: @Composable () -> Dp
    val roundedCornerShape: @Composable () -> Dp
    val textStyle: @Composable () -> TextStyle
}
```

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/ButtonStyleApearance.kt:
```diff
@@ -0,0 +1,64 @@
+            else -> UiKitTheme.colors.primary.copy(alpha = 0.8f)
+        }
+    },
+    override val size: @Composable () -> DpSize = { DpSize(width = 270.dp, height = 56.dp) },
```
Перепроверь размеры и скругление по фигме и обнови тут значения. Не совпадают

ilhan76 replied:
<img width="230" alt="Screenshot 2025-03-18 at 14 34 24" src="https://github.com/user-attachments/assets/7acab00d-c656-46f8-a9bd-ea25fccd0e20" />

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/ButtonStyleApearance.kt:
```diff
@@ -0,0 +1,64 @@
+    }
+) : IntButtonSize, IntButtonAppearance
+
+data class ButtonSizeMedium(
```
Я бы сказал, это уже Big кнопка

ilhan76 commented:
И два момента общих

- Обнови описание пра
- Скрин все еще лежит в проекте, надо бы удалить

AsWiMoN committed:
Delete Screenshot directory

Commit:
исправил:
имя композабл,
имена кнопки с медиум на биг,
размеры как в фигме,
изменил в Size параметр height,
проверил автоформат в файлах,
TextStyle берется из темы

Commit:
Merge remote-tracking branch 'origin/create-actor-screen-add-button' into create-actor-screen-add-button

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/RegularButton.kt:
```diff
@@ -0,0 +1,85 @@
+    appearance: RegularButtonAppearance = RegularButtonAppearances.Primary,
+    size: RegularButtonSize = RegularButtonSizes.Medium,
+    isEnabled: Boolean = true,
+    text: String = stringResource(id = R.string.delete_button),
```
Не должно быть значения по умолчанию. Этот текст в превью можно передать

ilhan76 replied:
Необязательным параметром и в самый верх функции

```
@Composable
fun RegularButton(
    text: String,
    modifier: Modifier = Modifier,
    ...
)
```

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/RegularButton.kt:
```diff
@@ -0,0 +1,85 @@
+    size: RegularButtonSize = RegularButtonSizes.Medium,
+    isEnabled: Boolean = true,
+    text: String = stringResource(id = R.string.delete_button),
+    onClick: () -> Unit,
```
А вот тут должно быть, как раз
    onClick: () -> Unit = {}

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/RegularButton.kt:
```diff
@@ -0,0 +1,85 @@
+        modifier = modifier
+            .fillMaxWidth()
+            .size(size = size.size())
+            .clip(RoundedCornerShape((cornerRadius)))
```
В Appearance выносим всю форму, чтобы тут сделать просто 

.clip(size.shape())

ilhan76 replied:
И название сделать более абстратным. roundedCornerShape -> backgroundShape, например. Чтобы если появится вдруг кнопка без скруглений, не было путаницы

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/RegularButtonSize.kt:
```diff
@@ -0,0 +1,35 @@
+}
+
+data class RegularButtonSizeSmall(
+    override val size: @Composable () -> Dp = { 26.dp},
```
А где в фигме есть такая кнопка? Вижу только вот такую. Вроде бы она самая маленькая

<img width="575" alt="Screenshot 2025-03-20 at 00 42 02" src="https://github.com/user-attachments/assets/ce839965-81fc-4a4c-92e0-c2c4e06562c0" />

ilhan76 commented:
Поправь описание согласно доки отсюда, пж) https://github.com/ilhan76/Roll-Initiative/pull/5/files

Commit:
исправил замечания про shape, размер кнопки

ilhan76 commented on app/src/main/java/ru/kudashov/rollinitiative/ui/element/button/regular/RegularButton.kt:
```diff
@@ -0,0 +1,83 @@
+    Box(
+        modifier = modifier
+            .fillMaxWidth()
+            .size(size = size.size())
```
Здесь должно быть .height(size.height())

Commit:
исправил замечания size.size на size.height(), обрезал через background
